common/src/main/scala/explore/utils/package.scala

```diff
@@ -254,3 +258,54 @@ extension [F[_], O](s: Stream[F, O])
         }.stream
       }
     }
+
+/**
+ * Produces a `Pipe` that evaluates an effect for each element in the input stream and emits its
+ * result, canceling previous effects that are still running with the same `key`.
+ */
+def keyedSwitchEvalMap[F[_]: Concurrent, I, O, K](
+  key: I => K,
+  f:   I => F[O]
+): Pipe[F, I, O] = {
+  def go(
+    stream: Stream[F, Either[Option[I], (K, O)]], // input? | key -> output
+    fibers: Map[K, Fiber[F, Throwable, Unit]],
```

Comment:
> There is always at max 1 running effect per key, and that is cancelled when pulling the new element.

Yes, that is the case.

> I think this could be simplified by using a [HotSwap](https://typelevel.org/cats-effect/docs/std/hotswap) instead of managing fibers directly?

`HotSwap` acts on `Resource`s. Do you mean running the fibers as `Resource`s with allocation as `.start` and releasing as `.cancel`? We would have to store a `HotSwap` for each key in a `Map` instead of `fibers`. That would probably work, but I'm not sure it's a simplification.

Reply:
Hmm, agreed this is probably the simpler solution